contributors/design-proposals/lifecycle-hooks.md

```diff
@@ -0,0 +1,164 @@
+    // Can have multiple hooks at the same ProgressPoint; order independent
+    Hooks []LifecycleHook
+}
+
+type LifecycleHook struct {
+    // Unique name
+    Name string
+    
+    // ProgressPoint specifies the point during update when the hook should be triggered.
+    // Accepts both the number of new pods available or a percentage value representing the ratio
+    // of new.availableReplicas to the declared replicas. In case of getting a percentage
+    // it will try to reach the exact or the closest possible point right after it,
+    // trigger the job, wait for it to complete and then continue. 
+    // If such situation shall occur that two different ProgressPoints should be reached at
+    // the same time, all the hooks for an earlier ProgressPoint will be ran (and finished)
+    // before any later one.
+    ProgressPoint intstr.IntOrString
+    
+    // "Abort" - Failure for this hook is fatal, deployment is considered failed and should be rollbacked.
+    // "Ignore" - Even if the hook fails deployment continues rolling out new version.
+    FailurePolicy string
+    
+    JobSpec v1.JobSpec
+}
+
+type LifecycleHookStatus struct {
+    // Name of the hook
+    Name string
+    
+    ProgressPoint intstr.IntOrString
+    
+    // States: "Running", "Succeeded", "Failed" 
+    State string
+    
+    // Reference to locate the Job created when executing the hook
+    JobRef LocalObjectReference
+}
+
+type LifecycleHookRevisionStatus struct {
+    // Revision of the object that this hook was run for
+    Revision string
+    
+    LifecycleHookStatuses []LifecycleHookStatus
+}
+```
+
+### Affected Objects
+As of now the lifecycle hooks are relevant for Deployments, StatefulSets and DaemonSets. 
+They will require new optional field to be able to reference LifecycleTemplate and extending its status by []LifecycleHookRevisionStatus. 
+Also the controller will need to be slightly adjusted to scale in appropriate chunks and trigger the hooks.
+
+```go
+type DeploymentSpec struct {
+    // ...
+    // Addition
+    LifecycleTemplate *ObjectReference
+}
+
+type DeploymentStatus struct {
+    // ...
+    // Addition
+    LifecycleHookRevisionStatuses []LifecycleHookRevisionStatus
+}
+```
+
+```go
+type StatefulSetSpec struct {
+    // ...
+    // Addition
+    LifecycleTemplate *ObjectReference
+}
+
+type StatefulSetStatus struct {
+    // ...
+    // Addition
+    LifecycleHookRevisionStatuses []LifecycleHookRevisionStatus
+}
+
+```
+
+```go
+type DaemonSetSpec struct {
+    // ...
+    // Addition
+    LifecycleTemplate *ObjectReference
+}
+
+type DaemonSetStatus struct {
+    // ...
+    // Addition
+    LifecycleHookRevisionStatuses []LifecycleHookRevisionStatus
+}
+```
+
+## Algorithm
+If there is a LifecycleTemplate referenced from an object:
+
+1. Calculate next partition point to reach the closest lifecycle hook progress point and scale replicas in update appropriately. If there is no hook remaining, GOTO 5.
```

Comment:
Some concrete examples for 3:

1. running a database migration - should be reentrant
2. clearing a cache - should be reentrant
3. notifying an external system of our current location - should be reentrant

I think at least once is probably ok, but I don't know that all users would expect at-least-once.  Similar to daemonset discussion - will users naturally assume these are at-most-once, and if so, will they experience catastrophic failure due to it?

Reply:
Of the examples from the previous comment, I think all of them imply the controller MUST block until the job completes.  Not running a database migration could be catastrophic.

Reply:
> 1. Does the controller block until the Job completes?
Yes. (Although there is always possibility of adding an option to customize it.) 

> 3. Do we require the Jobs to be idempotent for safety. I think it will be difficult to ensure at most once execution. We could probably ensure at least once.
User will get whatever guarantees Kubernetes Job gives him. (But it should be idempotent.)

> 4. If we are using available replicas as the trigger, what happens if the number of available replicas decreases during execution of the Job?
Nothing, once we reach certain progress point we trigger the hook and never go back.

Reply:
need to think about details of rollback in case of 2. more but it should be based on revision

Reply:
>How do you determine if a Job exists for a particular execution of the rolling update? That is, it seems to me that the execution of a Job is tied to the roll out of a specific revision, how do we track the lifecycle of the an individual Job for each roll out and rollback of a particular revision?

Would this be the same mechanism we use to determine whether a replicaset exists for a given deployment? That is, a deterministically named job based on the pod template spec hash.

Reply:
I think we could associate all the hooks (Jobs) by setting controllerRef to particular RS. If we decide to run hooks for rollbacks, those could be distinguished by label. 

We would have probably done it either way for purposes of adoption and garbage collection.